internal/query/rolebindings.go

```diff
@@ -0,0 +1,844 @@
+		update, err := e.rolebindingRelationshipUpdateForSubject(id, rb.ID.String(), pb.RelationshipUpdate_OPERATION_DELETE)
+		if err != nil {
+			span.RecordError(err)
+			span.SetStatus(codes.Error, err.Error())
+			logRollbackErr(e.logger, e.store.RollbackContext(dbCtx))
+
+			return types.RoleBinding{}, err
+		}
+
+		updates[i] = update
+		i++
+	}
+
+	if err := e.applyUpdates(ctx, updates); err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+		logRollbackErr(e.logger, e.store.RollbackContext(dbCtx))
+
+		return types.RoleBinding{}, err
+	}
+
+	// 3. update the role-binding in the database to record latest `updatedBy` and `updatedAt`
+	rbFromDB, err := e.store.UpdateRoleBinding(dbCtx, actor.ID, rb.ID)
+	if err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+		logRollbackErr(e.logger, e.store.RollbackContext(dbCtx))
+		logRollbackErr(e.logger, e.rollbackUpdates(ctx, updates))
+	}
+
+	if err := e.store.CommitContext(dbCtx); err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+		logRollbackErr(e.logger, e.store.RollbackContext(dbCtx))
+		logRollbackErr(e.logger, e.rollbackUpdates(ctx, updates))
+
+		return types.RoleBinding{}, err
+	}
+
+	rolebinding.Subjects = subjects
+	rolebinding.UpdatedAt = rbFromDB.UpdatedAt
+	rolebinding.UpdatedBy = rbFromDB.UpdatedBy
+
+	return rolebinding, nil
+}
+
+func (e *engine) GetRoleBindingResource(ctx context.Context, rb types.Resource) (types.Resource, error) {
+	rbFromDB, err := e.store.GetRoleBindingByID(ctx, rb.ID)
+	if err != nil {
+		if errors.Is(err, storage.ErrRoleBindingNotFound) {
+			err = fmt.Errorf("%w: %s", ErrRoleBindingNotFound, err)
+		}
+
+		return types.Resource{}, err
+	}
+
+	return e.NewResourceFromID(rbFromDB.ResourceID)
+}
+
+// isRoleBindable checks if a role is available for a resource. a role is not
+// available to a resource if its owner is not associated with the resource
+// in any way.
+func (e *engine) isRoleBindable(ctx context.Context, role, res types.Resource) error {
+	req := &pb.CheckPermissionRequest{
+		Resource: &pb.ObjectReference{
+			ObjectType: e.namespaced(res.Type),
+			ObjectId:   res.ID.String(),
+		},
+		Subject: &pb.SubjectReference{
+			Object: &pb.ObjectReference{
+				ObjectType: e.namespaced(e.rbac.RoleResource.Name),
+				ObjectId:   role.ID.String(),
+			},
+		},
+		Permission: iapl.AvailableRolesList,
+		Consistency: &pb.Consistency{
+			Requirement: &pb.Consistency_FullyConsistent{FullyConsistent: true},
+		},
+	}
+
+	err := e.checkPermission(ctx, req)
+
+	switch {
+	case err == nil:
+		return nil
+	case errors.Is(err, ErrActionNotAssigned):
+		return fmt.Errorf("%w: role: %s is not available for resource: %s", ErrRoleNotFound, role.ID, res.ID)
+	default:
+		return err
+	}
+}
+
+// deleteRoleBindingsForRole deletes all role-binding relationships with a given role.
+func (e *engine) deleteRoleBindingsForRole(ctx context.Context, roleResource types.Resource) error {
```

Comment:
I already noted this, but I'm not sure this is necessary. It looks like it has the potential to take a lot of time, and hard deleting every binding for a role seems like a pretty risky and destructive action.

In general, stopping users from deleting something that might have cascading effects seems more reasonable.